File: solutions/Reverse Vowels of a String/Reverse Vowels of a String.ts
/**
    STRATEGY:
        use 2 pointers, front and back.
        when both are at vowels, swap
    
    Notes:
        preserve casing 
 */
function reverseVowels(s: string): string {
    const sArr: string[] = s.split("");
    
    let front: number = 0;
    let back: number = s.length-1;

    while (front < back) {
        // move pointers to vowels
        if (!isVowel(sArr[front])) front++;
        if (!isVowel(sArr[back])) back--;
        
        // when both pointers are at vowels
        if (isVowel(sArr[front]) && isVowel(sArr[back])) {
            const temp: string = sArr[front];
            sArr[front] = sArr[back];
            sArr[back] = temp;
            front++;
            back--;
        }
    }

    return sArr.join("");
};

function isVowel(ch: string) {
    return ['a','e','i','o','u'].includes(ch.toLowerCase());
}

File: solutions/Reverse Vowels of a String/Reverse Vowels of a String.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = {'a','e','i','o','u'} # set
        
        sArr = [ch for ch in s]
        l, r = [0, len(s)-1]

        while (l < r):
            # point both pointers to vowels
            while ( l < r and sArr[l].lower() not in vowels):
                l+=1

            while ( l < r and sArr[r].lower() not in vowels ):
                r-=1

            # swap chars
            temp = sArr[l]
            sArr[l] = sArr[r]
            sArr[r] = temp

            # move pointers after swap
            l+=1
            r-=1

        return "".join(sArr)
